README.md
# pyTivo

## Description

pyTivo lets you stream most videos from your PC to your unhacked tivo.

pyTivo is both an [HMO][HMO spec] and GoBack server. Similar to [TiVo Desktop][],
pyTivo loads many standard video compression codecs and outputs mpeg2 (or in some
cases, h.264) video to the TiVo. However, pyTivo is able to load many more file
types than TiVo Desktop.

pyTivo is in no way affiliated with [TiVo, Inc][TiVo]. 

The pyTivo information here and more is available on the [pyTivo Wiki][] hosted on Sourceforge.

## Requirements

OS = Anything that will run python and ffmpeg, which I think is
anything. Known to work on Linux, Mac OS X and Windows.

Python - http://www.python.org/download/

- You need python version >= 2.5 and < 3.0 i.e. get the latest version 2 Python.

pywin32 (only to install as a service) -
http://sourceforge.net/project/showfiles.php?group_id=78018&package_id=79063
- Windows users only and only if you intend to install as a service

## Usage

You need to edit pyTivo.conf in 3 places

1. ffmpeg=
2. [&lt;name of share>]
3. path=

`ffmpeg` ([download][ffmpeg download]) should be the full path to ffmpeg including filename.
`path` is the absolute path to your media, which may be a network share. See the comments
in the sample `pyTivo.conf.dist` file.

run `pyTivo.py`

### To install as a service in Windows

run `pyTivoService.py --startup auto install`

### To remove service

run `pyTivoService.py remove`

## Additional Help

1. [Frequently Asked Questions (FAQ)][pyTivo FAQ]
1. [pyTivo Forum][]
1. [pyTiVo thread][] at TiVo Community Forum 

## Notes
pyTivo was created by Jason Michalski ("armooo"). Contributors include
Kevin R. Keegan, William McBrine, and Terry Mound ("wgw").

[HMO spec]: <http://tivopod.sourceforge.net/tivohomemedia.pdf> "TiVo Home Media Option specification"
[TiVo Desktop]: <https://support.tivo.com/articles/Installation_Setup_Configuration/TiVo-Desktop-Desktop-Plus-for-PC-Installation-and-Use> "TiVo Desktop support"
[TiVo]: <https://www.tivo.com/> "TiVo website"
[pyTivo Wiki]: <https://pytivo.sourceforge.io/wiki/index.php/PyTivo> "pyTivo Wiki"
[ffmpeg download]: <https://ffmpeg.org/download.html> "Download FFmpeg"
[pyTivo FAQ]: <https://pytivo.sourceforge.io/wiki/index.php/Frequently_Asked_Questions> "pytivo FAQ"
[pyTivo Forum]: <https://pytivo.sourceforge.io/forum/> "pyTivo Forum"
[pyTiVo thread]: <http://www.tivocommunity.com/tivo-vb/showthread.php?t=328459> "pyTiVo thread on TiVo Community Forum"

## Installing on Windows ##

Install [Python 3][py3-download]
* 3.6.2 is the latest version as of 9/13/2017
* it may be installed alongside of Python 2 and the installed [Python Launcher][]
  will help run the correct version.

The netifaces dependency requires Microsoft Visual C++ 14.0. Get it with "Microsoft Visual C++ Build Tools": http://landinghub.visualstudio.com/visual-cpp-build-tools

Clone this [repository][py3Tivo-repo] (e.g. into c:\py3Tivo)
Open an Administrator Command Prompt (if you installed Python 3 into Program Files) and install
the required python 3 packages (this should work even with both Python 2.7 and Python 3.x installed).

    c:
    cd \py3Tivo
    py -3 -m pip install -r requirements.txt

[py3-download]: <https://www.python.org/downloads/> (Python 3 Downloads)
[Python Launcher]: <https://docs.python.org/3/using/windows.html#python-launcher-for-windows>
[py3Tivo-repo]: <https://github.com/mlippert/pytivo> "py3Tivo git repository"


## Setting up development on Ubuntu 16.04 linux

### python3

From DigitalOcean's [page on setting up Python 3](https://www.digitalocean.com/community/tutorials/how-to-install-python-3-and-set-up-a-programming-environment-on-an-ubuntu-16-04-server):

Install these apt packages:

- python3-dev
- python3-venv
- build-essential
- libssl-dev
- libffi-dev

I had listed these, but I don't think they're needed (at least on Linux Mint 18.3)

- python3-pip (already installed?)
- pylint3 (install via pip instead)

Create a Python 3 virtual environment for py3tivo (see DigitalOcean's page)

    cd <pytivo repository root> and create the virtual env directory in there
    python3 -m venv venv
    pip install --upgrade pip
    pip install pylint


Install these python packages (with pip):

- zeroconf
- mutagen
- cheetah3
- pytz
- tzlocal

The required python packages are in `requirements.txt` and may be installed by running:

    pip install -r requirements.txt

## Development TODO

### Python 3 conversion (Done)

- Start from latest code in wmcbrine's pytivo repository
- run 2to3 over all files
- replace local Cheetah copy by using `pip install Cheetah3`
- replace local mutagen copy by using `pip install mutagen`
- replace local zeroconf copy by using `pip install zeroconf`
- start fixing Python 3 conversion issues that 2to3 didn't catch as they manifest
    - many str/bytes issues
    - seems Cheetah3 raises `NameMapper.NotFound: cannot find 'size' while searching for 'video.size'`
      while the old one silently replaced the missing attribute with nothing


Most important functionality (for me)

- togo: be able to download videos in TS from TiVos
- pull: Pull video from pyTivo share to TiVo

I've never used the music & picture sharing functionality so I'm not even sure at
this point how to test them.

Next:

- (**Done** - 1.6.7) Merge in desired changes from Dan203's [fork][dan203 pytivo]
- Refactor togo plugin for clarity and thread safety
    - (**Done**) extract download thread functionity into separate module/class
    - NPL (should be looked at, may or may not need changes)
- Add something like kmttg's naming for saving togo files
    - (**Done**) Create new ShowInfo class to encapsulate a show's metadata
    - Some metadata is no longer being supplied from the TiVo, may have to get it from [mind][tivo mind thread]
- [Pre-compile][cheetah compile] the Cheetah templates [sample makefile][cheetah makefile tip]
- see if it makes sense to replace `lrucache.py` with [`@functools.lru_cache`][lru_cache] decorator
- see if it makes sense to replace [`urllib.request`][py3 urllib.request] with the
  [Requests package][py3 requests pkg].
- update the use of [configparser][py3 configparser] because [Legacy API Ex.][py3 config legacy]
  says "...mapping protocol access is preferred for new projects."
- use subprocess.run instead of Popen where it makes sense.
- consistently return a ffmpeg options in lists not strings in transcode.py
- enhance web UI (not even sure what that means yet)

[cheetah makefile tip]: <https://pythonhosted.org/Cheetah/users_guide/tipsAndTricks.html#makefiles>
[cheetah compile]: <https://pythonhosted.org/Cheetah/recipes/precompiled.html>
[dan203 pytivo]: <https://github.com/Dan203/pytivo> "Dan Haddix's pytivo fork on github"
[lru_cache]: <https://docs.python.org/3/library/functools.html>
[py3 urllib.request]: <https://docs.python.org/3/library/urllib.request.html?highlight=request#module-urllib.request>
[py3 requests pkg]: <http://docs.python-requests.org/en/master/>
[py3 configparser]: <https://docs.python.org/3/library/configparser.html?highlight=configparser#module-configparser>
[py3 config legacy]: <https://docs.python.org/3/library/configparser.html?highlight=configparser#legacy-api-examples>
[tivo mind thread]: <https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=4&cad=rja&uact=8&ved=0ahUKEwi97OrQk6PVAhVFFT4KHZg7ASkQFgg4MAM&url=http%3A%2F%2Fwww.tivocommunity.com%2Fcommunity%2Findex.php%3Fthreads%2Fwhich-tivo-api-for-show-list-and-forced-playing-on-tivo.539075%2F&usg=AFQjCNHBHt_ng1m_03xDNk_04hj1ewTTtA>

CHANGELOG.md
# Change Log

## [2.6.2] - 2021-02-22

update for python 3.9 and latest dependency package versions

### Fixed

- pylint removed bad-whitespace error. see [whatsnew 2.6](https://pylint.pycqa.org/en/latest/whatsnew/2.6.html)
- updating [zeroconf](https://github.com/jstasiak/python-zeroconf) (also see
  [latest docs](https://python-zeroconf.readthedocs.io/en/latest/api.html))
  from 0.24 to 0.28 required several changes in beacon.py
    - ServiceInfo address member was deprecated and then removed, addresses should now be used
    - ServiceInfo ctor argument order changed, and address was replaced by addresses
    - ServiceListeners will soon require an update_service method although it may do nothing
- python 3.9 plistlib module has removed the old API. It's used by metadata in a section of code
  that I do not think has ever been tested by me after updated from python 2 to 3. I've updated
  the code as I believe is needed, but it is *still* untested.
  see [python docs 3.5 plistlib](https://docs.python.org/3.5/library/plistlib.html)
- python 3.9 xml.xmlparser expat is giving errors (I'm not totally sure these errors didn't
  exist before). Some googling also made me believe in this instance the pylint errors may
  be false positives.
    - `metadata.py:608:23: E1101: Instance of 'module' has no 'codes' member (no-member)`
    - `metadata.py:608:42: E1101: Instance of 'module' has no 'XML_ERROR_INVALID_TOKEN' member (no-member)`
- fixed a typo error in plugins/video/video.py


## [2.6.1] - 2020-02-23

### Fixed

- Fix lrucache for PEP 479 implemented in python 3.6+. pyTivo now works with python 3.8


## [2.6.0] - 2018-06-10

### Changed

- Enhanced syncerr yaml report file by adding tivoName, attempt transfer info
  and error startMB.


## [2.5.1] - 2018-02-13

### Fixed

- fix crash getting NPL of a different TiVo than the one with active downloads.


## [2.5.0] - 2018-02-01

### Changed

- add new _ts_error_mode_ value **all** that saves all togo download attempts
- refactor the pyTivo version references in the code and add the version to
  the info page
- Update dependecies
    - mutagen (1.40)


## [2.4.0] - 2018-01-26

### Changed

- config fields for customized togo file naming; _episode_fn_ &  _movie_fn_,
  see [togo/fn_fields.md](./plugins/togo/fn_fields.md) for more info
- change logging priority of httpserver requests from info to debug to reduce
  noise when sending info priority messages to the console


## [2.3.0] - 2018-01-09

Some testing on Windows 7 in addition to Linux

### Changed

- Write a yaml sync error log file for every togo download
- Default to using transport stream downloads
- Update dependecies
    - mutagen (1.39)
    - pytz (2017.3)
    - tzlocal (1.5.1)
- Add a section to the Readme about installing on MS Windows

### Added

- Add some TiVo documentation so it's available in the future if needed.
- Add some development features like a Makefile to help running pylint and
  other development tasks


## [2.2.1] - 2017-07-30

### Fixed

- fix exception putting attempt number in filename
- improve logging


## [2.2.0] - 2017-07-29

### Changed

- Improve the logging during togo downloads
- Implement new ShowInfo class to encapsulate most of the show metadata
- Refactor download thread functionality
- Update dependecies, add tzlocal
    - mutagen (1.38)
    - zeroconf (0.19.1)
    - tzlocal (1.4)


## [2.1.0] - 2017-07-18

### Changed

- Incorporate Dan203's changes for his version PyTivo up through 1.6.7 except the
  pyInstaller and pyTivoTray which were more specifically for PyTivoDesktop and
  Windows & Mac not Linux.
    - Check transport stream (ts) downloads for sync errors
    - TivoDesktop togo file naming
    - support using tivolibre to decode the downloaded .tivo file
- moved togo settings to their own section in pytivo.conf ([togo])

### Fixed

- Implement thread safety for the togo download threads
- Log not finding a binary (e.g. tivodecode) only once per run

### Added

- new dependency
    - pytz (2017.2)


## [2.0.0] - 2017-06-20 (forked to https://github.com/mlippert/pytivo)

Tested only on Linux

### Changed

- Converted from Python 2 to **Python 3**
- Remove local copies of packages, use latest versions from pip
    - zeroconf (0.19.0)
    - mutagen (1.37)
    - cheetah3 (3.0.0)
- Format README with markdown
- Various info added to the README
- Prefix the episode title w/ the episode number when shown in a folder on the TiVo

### Added

- requirements.txt file for use w/ pip to install dependencies


## [1.6.0] - prior to 2017-01-01 (from https://github.com/wmcbrine/pytivo)

### Added

- All functionality of pyTivo up until this time. (I am not aware of any distinct
  versions or releases prior to this time. -mjl)


config.py
import getopt
import logging
import logging.config
import os
import re
import socket
import sys
import uuid
import configparser
from configparser import NoOptionError
from functools import reduce


# Any configuration section whose name is not special is considered a share section
special_section_names = ('Server',      # for main pytivo settings
                         'togo',        # for togo settings
                         'loggers',     # for advanced logging (see logging.config)
                         'handlers',    #  "
                         'formatters',  #  "
                        )
special_section_prefixes = ('_tivo_',       # for override setting for a tivo w/ a specific tsn
                            'logger_',      # for advanced logging (see logging.config)
                            'handler_',     #  "
                            'formatter_',   #  "
                           )

SCRIPTDIR = os.path.dirname(__file__)

# global variables
tivos = {}
tivos_found = False
guid = uuid.uuid4()
config_files = ['/etc/pyTivo.conf', os.path.join(SCRIPTDIR, 'pyTivo.conf')]
configs_found = False
config = None
bin_paths = {}

class Error(Exception):
    """Base class for exceptions in this module."""
    pass

class Bdict(dict):
    def getboolean(self, x):
        return self.get(x, 'False').lower() in ('1', 'yes', 'true', 'on')

def init(argv):
    global config_files

    try:
        opts, _ = getopt.getopt(argv, 'c:e:', ['config=', 'extraconf='])
    except getopt.GetoptError as msg:
        print(msg)

    for opt, value in opts:
        if opt in ('-c', '--config'):
            config_files = [value]
        elif opt in ('-e', '--extraconf'):
            config_files.append(value)

    reset()

def reset():
    global bin_paths
    global config
    global configs_found
    global tivos_found

    bin_paths = {}

    config = configparser.ConfigParser(interpolation=None)
    configs_found = config.read(config_files)
    if not configs_found:
        print(('WARNING: pyTivo.conf does not exist.\n'
               'Assuming default values.'))
        configs_found = config_files[-1:]

    for section in config.sections():
        if section.startswith('_tivo_'):
            tsn = section[6:]
            if tsn.upper() not in ['SD', 'HD', '4K']:
                tivos_found = True
                tivos[tsn] = Bdict(config.items(section))

    for section in ['Server', '_tivo_SD', '_tivo_HD', '_tivo_4K']:
        if not config.has_section(section):
            config.add_section(section)

def write():
    f = open(configs_found[-1], 'w')
    config.write(f)
    f.close()

def tivos_by_ip(tivoIP):
    """
    Get the tsn for a tivo with the given IPv4 address
    """
    for key, value in tivos.items():
        # some tivo entries may not have an 'address' if they were
        # created from a config _tivo_TSN section
        if 'address' in value and value['address'] == tivoIP:
            return key
    raise Error('No TiVo w/ IP:{} was found'.format(tivoIP))

def get_server(name, default=None):
    if config.has_option('Server', name):
        return config.get('Server', name)
    return default

def get_togo(name, default=None):
    if config.has_option('togo', name):
        return config.get('togo', name)

    # many togo options used to be in the server section with
    # the name prefixed w/ 'togo_', so check for those values
    # before returning the default
    return get_server('togo_{}'.format(name), default)

def getGUID():
    return str(guid)

def get_ip(tsn=None):
    try:
        assert tsn
        dest_ip = tivos[tsn]['address']
    except:
        dest_ip = '4.2.2.1'
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect((dest_ip, 123))
    return s.getsockname()[0]

def get_zc():
    opt = get_server('zeroconf', 'auto').lower()

    if opt == 'auto':
        for section in config.sections():
            if section.startswith('_tivo_'):
                if config.has_option(section, 'shares'):
                    logger = logging.getLogger('pyTivo.config')
                    logger.info('Shares security in use -- zeroconf disabled')
                    return False
    elif opt in ['false', 'no', 'off']:
        return False

    return True

def getBeaconAddresses():
    return get_server('beacon', '255.255.255.255')

def getPort():
    return get_server('port', '9032')

def get169Blacklist(tsn):  # tivo does not pad 16:9 video
    return tsn and not isHDtivo(tsn) and not get169Letterbox(tsn)
    # verified Blacklist Tivo's are ('130', '240', '540')
    # It is assumed all remaining non-HD and non-Letterbox tivos are Blacklist

def get169Letterbox(tsn):  # tivo pads 16:9 video for 4:3 display
    return tsn and tsn[:3] in ['649']

def get169Setting(tsn):
    if not tsn:
        return True

    tsnsect = '_tivo_' + tsn
    if config.has_section(tsnsect):
        if config.has_option(tsnsect, 'aspect169'):
            try:
                return config.getboolean(tsnsect, 'aspect169')
            except ValueError:
                pass

    if get169Blacklist(tsn) or get169Letterbox(tsn):
        return False

    return True

def getAllowedClients():
    return get_server('allowedips', '').split()

def getIsExternal(tsn):
    tsnsect = '_tivo_' + tsn
    if tsnsect in config.sections():
        if config.has_option(tsnsect, 'external'):
            try:
                return config.getboolean(tsnsect, 'external')
            except ValueError:
                pass

    return False

def isTsnInConfig(tsn):
    return ('_tivo_' + tsn) in config.sections()

def getShares(tsn=''):
    shares = [(section, Bdict(config.items(section)))
              for section in config.sections()
              if not (section.startswith(special_section_prefixes)
                      or section in special_section_names)
             ]

    tsnsect = '_tivo_' + tsn
    if config.has_section(tsnsect) and config.has_option(tsnsect, 'shares'):
        # clean up leading and trailing spaces & make sure ref is valid
        tsnshares = []
        for x in config.get(tsnsect, 'shares').split(','):
            y = x.strip()
            if config.has_section(y):
                tsnshares.append((y, Bdict(config.items(y))))
        shares = tsnshares

    shares.sort()

    if not config.getboolean('Server', 'nosettings', fallback=False):
        shares.append(('Settings', {'type': 'settings'}))
    if get_server('tivo_mak') and get_togo('path'):
        shares.append(('ToGo', {'type': 'togo'}))

    return shares

def getDebug():
    try:
        return config.getboolean('Server', 'debug')
    except:
        return False

def getOptres(tsn=None):
    try:
        return config.getboolean('_tivo_' + tsn, 'optres')
    except:
        try:
            return config.getboolean(get_section(tsn), 'optres')
        except:
            try:
                return config.getboolean('Server', 'optres')
            except:
                return False

def get_bin(fname):
    logger = logging.getLogger('pyTivo.config')

    if fname in bin_paths:
        return bin_paths[fname]

    if config.has_option('Server', fname):
        fpath = config.get('Server', fname)
        if os.path.exists(fpath) and os.path.isfile(fpath):
            bin_paths[fname] = fpath
            return fpath
        else:
            logger.error('Bad %s path: %s', fname, fpath)

    if sys.platform == 'win32':
        fext = '.exe'
    else:
        fext = ''

    for path in ([os.path.join(SCRIPTDIR, 'bin')] +
                 os.getenv('PATH').split(os.pathsep)):
        fpath = os.path.join(path, fname + fext)
        if os.path.exists(fpath) and os.path.isfile(fpath):
            bin_paths[fname] = fpath
            return fpath

    logger.warning('%s not found', fname)
    bin_paths[fname] = None
    return None

def getFFmpegWait():
    if config.has_option('Server', 'ffmpeg_wait'):
        return max(int(float(config.get('Server', 'ffmpeg_wait'))), 1)
    return 0

def getFFmpegPrams(tsn):
    return get_tsn('ffmpeg_pram', tsn, True)

def isHDtivo(tsn):  # TSNs of High Definition TiVos
    return bool(tsn and tsn[0] >= '6' and tsn[:3] != '649')

def is4Ktivo(tsn):  # TSNs of 4K TiVos
    return bool(tsn[:3] in ('849', '8F9'))

def get_ts_flag():
    return get_server('ts', 'auto').lower()

def is_ts_capable(tsn):  # tsn's of Tivos that support transport streams
    return bool(tsn and (tsn[0] >= '7' or tsn.startswith('663')))

def getValidWidths():
    return [1920, 1440, 1280, 720, 704, 544, 480, 352]

def getValidHeights():
    return [1080, 720, 480] # Technically 240 is also supported

# Return the number in list that is nearest to x
# if two values are equidistant, return the larger
def nearest(x, lst):
    return reduce(lambda a, b: closest(x, a, b), lst)

def closest(x, a, b):
    da = abs(x - a)
    db = abs(x - b)
    if da < db or (da == db and a > b):
        return a
    return b

def nearestTivoHeight(height):
    return nearest(height, getValidHeights())

def nearestTivoWidth(width):
    return nearest(width, getValidWidths())

def getTivoHeight(tsn):
    if is4Ktivo(tsn):
        return 2160
    return 1080 if isHDtivo(tsn) else 480

def getTivoWidth(tsn):
    if is4Ktivo(tsn):
        return 3840
    return 1920 if isHDtivo(tsn) else 544

def _trunc64(i):
    return max(int(strtod(i)) / 64000, 1) * 64

def getAudioBR(tsn=None):
    rate = get_tsn('audio_br', tsn)
    if not rate:
        rate = '448k'
    # convert to non-zero multiple of 64 to ensure ffmpeg compatibility
    # compare audio_br to max_audio_br and return lowest
    return str(min(_trunc64(rate), getMaxAudioBR(tsn))) + 'k'

def _k(i):
    return str(int(strtod(i)) / 1000) + 'k'

def getVideoBR(tsn=None):
    rate = get_tsn('video_br', tsn)
    if rate:
        return _k(rate)
    if is4Ktivo(tsn):
        return getMaxVideoBR(tsn)
    return '16384K' if isHDtivo(tsn) else '4096K'

def getMaxVideoBR(tsn=None):
    rate = get_tsn('max_video_br', tsn)
    if rate:
        return _k(rate)
    return '30000k'

def getBuffSize(tsn=None):
    size = get_tsn('bufsize', tsn)
    if size:
        return _k(size)
    if is4Ktivo(tsn):
        return '8192k'
    return '4096k' if isHDtivo(tsn) else '1024k'

def getMaxAudioBR(tsn=None):
    rate = get_tsn('max_audio_br', tsn)
    # convert to non-zero multiple of 64 for ffmpeg compatibility
    if rate:
        return _trunc64(rate)
    return 448

def get_section(tsn):
    if is4Ktivo(tsn):
        return '_tivo_4K'
    return '_tivo_HD' if isHDtivo(tsn) else '_tivo_SD'

def get_tsn(name, tsn=None, raw=False):
    try:
        return config.get('_tivo_' + tsn, name, raw=raw)
    except:
        try:
            return config.get(get_section(tsn), name, raw=raw)
        except:
            try:
                return config.get('Server', name, raw=raw)
            except:
                return None

# Parse a bitrate using the SI/IEEE suffix values as if by ffmpeg
# For example, 2K==2000, 2Ki==2048, 2MB==16000000, 2MiB==16777216
# Algorithm: http://svn.mplayerhq.hu/ffmpeg/trunk/libavcodec/eval.c
def strtod(value):
    prefixes = {'y': -24, 'z': -21, 'a': -18, 'f': -15, 'p': -12,
                'n': -9,  'u': -6,  'm': -3,  'c': -2,  'd': -1,
                'h': 2,   'k': 3,   'K': 3,   'M': 6,   'G': 9,
                'T': 12,  'P': 15,  'E': 18,  'Z': 21,  'Y': 24}
    p = re.compile(r'^(\d+)(?:([yzafpnumcdhkKMGTPEZY])(i)?)?([Bb])?$')
    m = p.match(value)
    if not m:
        raise SyntaxError('Invalid bit value syntax')
    (coef, prefix, power, byte) = m.groups()
    if prefix is None:
        value = float(coef)
    else:
        exponent = float(prefixes[prefix])
        if power == 'i':
            # Use powers of 2
            value = float(coef) * pow(2.0, exponent / 0.3)
        else:
            # Use powers of 10
            value = float(coef) * pow(10.0, exponent)
    if byte == 'B': # B == Byte, b == bit
        value *= 8
    return value

def init_logging():
    if     (config.has_section('loggers') and
            config.has_section('handlers') and
            config.has_section('formatters')):
        logging.config.fileConfig(config_files)
    elif getDebug():
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.INFO)
